Compilation errors - related to openssl version?

Good day, I'm still investigating this, but curious if you are aware of typecasting errors of this library and issues related to openssl version, possibly? 

I am able to build your library and use it just fine, but I'm trying to integrate this into a 3rd party system and the build is failing on travis there. I am currently not sure if their VM is using different versions of libraries, but its on ubuntu (as am I).

I'll be digging in more, but if the error looks familiar, would love a hint. Maybe [this fix](https://github.com/eidheim/Simple-Web-Server/issues/150) is related?

```
../third_party/jwt-cpp/include/jwt-cpp/jwt.h: In function ‘std::string jwt::helper::extract_pubkey_from_cert(const string&, const string&)’:
../third_party/jwt-cpp/include/jwt-cpp/jwt.h:77:87: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]
    std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf(certstr.data(), certstr.size()), BIO_free_all);
                                                                                       ^
In file included from /usr/include/openssl/evp.h:75:0,
                 from ../third_party/jwt-cpp/include/jwt-cpp/jwt.h:9,
                 from src/zm_crypt.h:27,
                 from /src/zm_user.cpp:32:
/usr/include/openssl/bio.h:668:6: error:   initializing argument 1 of ‘BIO* BIO_new_mem_buf(void*, int)’ [-fpermissive]
 BIO *BIO_new_mem_buf(void *buf, int len);
      ^
```

--- a/include/jwt-cpp/jwt.h
+++ b/include/jwt-cpp/jwt.h
@@ -74,13 +74,19 @@ namespace jwt {
 	namespace helper {
 		inline
 		std::string extract_pubkey_from_cert(const std::string& certstr, const std::string& pw = "") {
-			std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf((char *)certstr.data(), certstr.size()), BIO_free_all);
+			// TODO: Cannot find the exact version this change happended
+#if OPENSSL_VERSION_NUMBER <= 0x1000114fL
+			std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf(const_cast<char*>(certstr.data()), certstr.size()), BIO_free_all);
+#else
+			std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf(certstr.data(), certstr.size()), BIO_free_all);
+#endif
 			std::unique_ptr<BIO, decltype(&BIO_free_all)> keybio(BIO_new(BIO_s_mem()), BIO_free_all);
 
 			std::unique_ptr<X509, decltype(&X509_free)> cert(PEM_read_bio_X509(certbio.get(), NULL, [](char *buf, int size, int rwflag, void *userdata)->int{
-				auto mpw = reinterpret_cast<const std::string*>(userdata)->substr(0, size);
-				memcpy(buf, mpw.data(), mpw.size());
-				return mpw.size();
+				auto mpw = reinterpret_cast<const std::string*>(userdata);
+				size = static_cast<size_t>(size) > mpw->size() ? mpw->size() : size;
+				memcpy(buf, mpw->data(), size);
+				return size;
 			}, const_cast<void*>(reinterpret_cast<const void*>(&pw))), X509_free);
 			if (!cert) throw rsa_exception("Error loading cert into memory");
 			std::unique_ptr<EVP_PKEY, decltype(&EVP_PKEY_free)> key(X509_get_pubkey(cert.get()), EVP_PKEY_free);

--- a/include/jwt-cpp/jwt.h
+++ b/include/jwt-cpp/jwt.h
@@ -82,12 +82,7 @@ namespace jwt {
 #endif
 			std::unique_ptr<BIO, decltype(&BIO_free_all)> keybio(BIO_new(BIO_s_mem()), BIO_free_all);
 
-			std::unique_ptr<X509, decltype(&X509_free)> cert(PEM_read_bio_X509(certbio.get(), NULL, [](char *buf, int size, int rwflag, void *userdata)->int{
-				auto mpw = reinterpret_cast<const std::string*>(userdata);
-				size = static_cast<size_t>(size) > mpw->size() ? mpw->size() : size;
-				memcpy(buf, mpw->data(), size);
-				return size;
-			}, const_cast<void*>(reinterpret_cast<const void*>(&pw))), X509_free);
+			std::unique_ptr<X509, decltype(&X509_free)> cert(PEM_read_bio_X509(certbio.get(), nullptr, nullptr, const_cast<char*>(pw.c_str())), X509_free);
 			if (!cert) throw rsa_exception("Error loading cert into memory");
 			std::unique_ptr<EVP_PKEY, decltype(&EVP_PKEY_free)> key(X509_get_pubkey(cert.get()), EVP_PKEY_free);
 			if(!key) throw rsa_exception("Error getting public key from certificate");

--- a/include/jwt-cpp/jwt.h
+++ b/include/jwt-cpp/jwt.h
@@ -74,7 +74,7 @@ namespace jwt {
 	namespace helper {
 		inline
 		std::string extract_pubkey_from_cert(const std::string& certstr, const std::string& pw = "") {
-			std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf(certstr.data(), certstr.size()), BIO_free_all);
+			std::unique_ptr<BIO, decltype(&BIO_free_all)> certbio(BIO_new_mem_buf((char *)certstr.data(), certstr.size()), BIO_free_all);
 			std::unique_ptr<BIO, decltype(&BIO_free_all)> keybio(BIO_new(BIO_s_mem()), BIO_free_all);
 
 			std::unique_ptr<X509, decltype(&X509_free)> cert(PEM_read_bio_X509(certbio.get(), NULL, [](char *buf, int size, int rwflag, void *userdata)->int{

